src/crawlee/storages/_request_provider.py

```diff
@@ -55,6 +55,24 @@ async def mark_request_as_handled(self, request: Request) -> ProcessedRequest |
     async def get_handled_count(self) -> int:
         """Returns the number of handled requests."""
 
+    @abstractmethod
+    async def add_request(
+        self,
+        request: str | Request,
+        *,
+        forefront: bool = False,
+    ) -> ProcessedRequest:
+        """Add a single request to the provider and store it in underlying resource client.
+
+        Args:
+            request: The request object (or its string representation) to be added to the provider.
+            forefront: Determines whether the request should be added to the beginning (if True) or the end (if False)
+                of the provider.
+
+        Returns:
+            Information about the request addition to the provider.
+        """
+
```

Comment:
I think this was missing intentionally - the `RequestProvider` interface already has `add_requests_batched`, so what's the reason to also require a single-request version?

Reply:
both methods should be available, we do have `addRequests` in the js version on this level

https://github.com/apify/crawlee/blob/master/packages/core/src/storages/request_provider.ts#L107

in the long run (crawlee v4) we want to get rid of the duality between RQ v1/v2, so we might as well remove this abstraction. not sure how things are implemented in the python version.

Reply:
Here, `RequestProvider` is just an interface for "something that BasicCrawler can pull requests from". Hopefully that will also be the case in the JS version one day... But here and now, I'd just not try to restore the parity with JS.

Reply:
hmm, and we dont use the `add_request` method internally anywhere now? 

i would say it should be here, in the end users can work with the RQ directly, not just via crawler.

Reply:
I believe `RequestQueue` has `add_request` for convenience, but it's not required by the `RequestProvider` interface. So currently you can do `RequestsQueue.open().add_request()` and pass the same queue into the crawler via constructor. You can't do `crawler.get_request_provider().add_request()` though.

Reply:
Maybe we could just add a default implementation of `add_request` to `RequestProvider` that would use the batched version?

Reply:
@B4nan good point, I like that more.